File: reports/AccountReportPage.java
package com.afm.Reports.pages;

import java.util.concurrent.TimeUnit;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.PageFactory;
import com.afm.base.BaseLibrary;
import com.afm.constants.AfmCommonConstants;
import com.afm.constants.AfmNewOrderConstants;
import com.afm.helper.CommonActions;
import com.afm.helper.ReadConfig;
import com.afm.helper.WaitUtility;

public class AccountReportPage extends BaseLibrary {
	public static Logger logger = LogManager.getLogger(AccountReportPage.class);
	public String orderNumber;
	public String tripNumber;
	public String driverName;
	public int sizeNumber;
	public int randomNumber;
	public String name;
	public String carrierName;
	public String customerName;
	public String salesName;
	public String truckName;
	public String driverTwoName;
	public String trailerName;
	public String trailerReferenceName;
	public String driverTwoReferenceName;
	public String paymentReferenceNumber;
	public String truckReferenceName;
	public String driverReferenceName;
	public String vendorName;
	public String vendorNameRandomNum;
	ReadConfig readConfig = new ReadConfig();
	String env = readConfig.getPropertiesValue("baseURL");

	/**
	 * Constructor to initialize the PageFactory
	 */
	public AccountReportPage() {
		PageFactory.initElements(driver, this);
	}

	/**
	 * Reports-->Account-->Profit & Loss Report Create
	 */
	public boolean profitAndLossReport() {

		try {
			logger.info("Profit And Loss Report Create");
			String dropFutureDate = CommonActions.getFutureDate(AfmCommonConstants.FIVE).concat(" 12:00 AM");
			CommonActions.clickOnElement(AfmNewOrderConstants.REPORTS_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.ACCOUNT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.PROFIT_LOSS_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.PROFIT_ADVANCE_SEARCH_XPATH);
			CommonActions.selectByVisibleText(AfmNewOrderConstants.PROFIT_CURRENCY_XPATH,
					AfmNewOrderConstants.CURRENCY_VALUE_CAD);
			CommonActions.clickAndClearText(AfmNewOrderConstants.PROFIT_DATE_XPATH);
			CommonActions.clickAndEnterText(AfmNewOrderConstants.PROFIT_DATE_XPATH, dropFutureDate);
			boolean isSubmitButtonDisplayed = CommonActions.verifyElementDisplayed(AfmNewOrderConstants.SUBMIT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.SUBMIT_XPATH);
			return isSubmitButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Profit & Loss Report Print
	 */
	public boolean profitAndLossReportPrint() {

		try {
			logger.info("Profit And Loss Report Print");
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			CommonActions.clickOnElement(AfmNewOrderConstants.REPORT_PRINT_XPATH);
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			boolean isRefershButtonDisplayed = CommonActions
					.verifyElementDisplayed(AfmNewOrderConstants.REFERSH_REPORT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.REFERSH_REPORT_XPATH);
			captureScreen(driver, "profitAndLossReportPrint");
			return isRefershButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Profit And Loss Report Email
	 */
	public boolean profitAndLossReportEmail() {

		try {
			logger.info("Profit And Loss Report Email");
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			WebElement ele = driver.findElement(By.xpath("//*[contains(@onclick,'EmailAFMReports')]"));
			JavascriptExecutor js = (JavascriptExecutor) driver;
			driver.manage().timeouts().implicitlyWait(1, TimeUnit.MINUTES);
			js.executeScript("arguments[0].click();", ele);
			CommonActions.clickAndEnterText(AfmNewOrderConstants.EMAIL_TO_XPATH, AfmNewOrderConstants.EMAIL_TO_VALUE);
			boolean isEmailSendButtonDisplayed = CommonActions
					.verifyElementDisplayed(AfmNewOrderConstants.EMAIL_SEND_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.EMAIL_SEND_XPATH);
			captureScreen(driver, "profitAndLossReportEmail");
			WaitUtility.pause(AfmCommonConstants.THREE);
			return isEmailSendButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Cheque Register Report
	 */
	public boolean chequeRegisterReport() {

		try {
			logger.info("Cheque Register Report");
			CommonActions.clickOnElement(AfmNewOrderConstants.ACCOUNT_XPATH);
			String dropFutureDate = CommonActions.getFutureDate(AfmCommonConstants.FIVE).concat(" 12:00 AM");
			CommonActions.clickOnElement(AfmNewOrderConstants.CHEQUE_REGISTER_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.CHEQUE_REGISTER_ADVANCE_SEARCH_XPATH);
			CommonActions.clickAndClearText(AfmNewOrderConstants.CHEQUE_REGISTER_DATE_XPATH);
			CommonActions.clickAndEnterText(AfmNewOrderConstants.CHEQUE_REGISTER_DATE_XPATH, dropFutureDate);
			boolean isSubmitButtonDisplayed = CommonActions.verifyElementDisplayed(AfmNewOrderConstants.SUBMIT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.SUBMIT_XPATH);
			return isSubmitButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Cheque Register Report Print
	 */
	public boolean chequeRegisterReportPrint() {

		try {
			logger.info("Profit And Loss Report Print");
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			CommonActions.clickOnElement(AfmNewOrderConstants.REPORT_PRINT_XPATH);
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			boolean isRefershButtonDisplayed = CommonActions
					.verifyElementDisplayed(AfmNewOrderConstants.REFERSH_REPORT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.REFERSH_REPORT_XPATH);
			captureScreen(driver, "chequeRegisterReportPrint");
			return isRefershButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Profit And Loss Report Email
	 */
	public boolean chequeRegisterReportEmail() {

		try {
			logger.info("Cheque Register Report Email");
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			WebElement ele = driver.findElement(By.xpath("//*[contains(@onclick,'EmailAFMReports')]"));
			JavascriptExecutor js = (JavascriptExecutor) driver;
			driver.manage().timeouts().implicitlyWait(1, TimeUnit.MINUTES);
			js.executeScript("arguments[0].click();", ele);
			CommonActions.clickAndEnterText(AfmNewOrderConstants.EMAIL_TO_XPATH, AfmNewOrderConstants.EMAIL_TO_VALUE);
			boolean isEmailSendButtonDisplayed = CommonActions
					.verifyElementDisplayed(AfmNewOrderConstants.EMAIL_SEND_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.EMAIL_SEND_XPATH);
			captureScreen(driver, "chequeRegisterReportEmail");
			WaitUtility.pause(AfmCommonConstants.THREE);
			return isEmailSendButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Bank Statement Report
	 */
	public boolean bankStatementReport() {

		try {
			logger.info("Bank Statement Report");
			String dropFutureDate = CommonActions.getFutureDate(AfmCommonConstants.FIVE).concat(" 12:00 AM");
			CommonActions.clickOnElement(AfmNewOrderConstants.ACCOUNT_XPATH);
			boolean isBankStatementButtonDisplayed = CommonActions
					.verifyElementDisplayed(AfmNewOrderConstants.BANK_STATEMENT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.BANK_STATEMENT_XPATH);
			CommonActions.selectByVisibleText(AfmNewOrderConstants.BANK_ACCOUNT_XPATH, carrierName);
			CommonActions.clickAndClearText(AfmNewOrderConstants.BANK_STATEMENT_DATE_XPATH);
			CommonActions.clickAndEnterText(AfmNewOrderConstants.BANK_STATEMENT_DATE_XPATH, dropFutureDate);
			boolean isSubmitButtonDisplayed = CommonActions.verifyElementDisplayed(AfmNewOrderConstants.SUBMIT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.SUBMIT_XPATH);
			captureScreen(driver, "bankStatementReport");
			return isBankStatementButtonDisplayed && isSubmitButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Bank Statement Report Print
	 */
	public boolean bankStatementReportPrint() {

		try {
			logger.info("Bank Statement Report Print");
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			CommonActions.clickOnElement(AfmNewOrderConstants.REPORT_PRINT_XPATH);
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			boolean isRefershButtonDisplayed = CommonActions
					.verifyElementDisplayed(AfmNewOrderConstants.REFERSH_REPORT_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.REFERSH_REPORT_XPATH);
			captureScreen(driver, "bankStatementReportPrint");
			return isRefershButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}

	/**
	 * Reports-->Account-->Bank Statement Report Email
	 */
	public boolean bankStatementReportEmail() {

		try {
			logger.info("Bank Statement Report Email");
			driver.manage().timeouts().implicitlyWait(2, TimeUnit.MINUTES);
			WebElement ele = driver.findElement(By.xpath("//*[contains(@onclick,'EmailAFMReports')]"));
			JavascriptExecutor js = (JavascriptExecutor) driver;
			driver.manage().timeouts().implicitlyWait(1, TimeUnit.MINUTES);
			js.executeScript("arguments[0].click();", ele);
			CommonActions.clickAndEnterText(AfmNewOrderConstants.EMAIL_TO_XPATH, AfmNewOrderConstants.EMAIL_TO_VALUE);
			boolean isEmailSendButtonDisplayed = CommonActions
					.verifyElementDisplayed(AfmNewOrderConstants.EMAIL_SEND_XPATH);
			CommonActions.clickOnElement(AfmNewOrderConstants.EMAIL_SEND_XPATH);
			captureScreen(driver, "bankStatementReportEmail");
			WaitUtility.pause(AfmCommonConstants.THREE);
			return isEmailSendButtonDisplayed;
		} catch (Exception e) {
			return false;
		}
	}
}

File: reports/AccountReportPage.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait, Select
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import Exception as SeleniumException
from selenium.webdriver.common.action_chains import ActionChains
from datetime import datetime, timedelta
import time


class AccountReportPage:
    def __init__(self, driver):
        self.driver = driver
        self.wait = WebDriverWait(driver, 20)

        # Replace these with your actual locators from constants
        self.REPORTS_XPATH = "your_reports_xpath"
        self.ACCOUNT_XPATH = "your_account_xpath"
        self.PROFIT_LOSS_XPATH = "your_profit_loss_xpath"
        self.PROFIT_ADVANCE_SEARCH_XPATH = "your_profit_advance_search_xpath"
        self.PROFIT_CURRENCY_XPATH = "your_profit_currency_xpath"
        self.PROFIT_DATE_XPATH = "your_profit_date_xpath"
        self.SUBMIT_XPATH = "your_submit_xpath"
        self.REPORT_PRINT_XPATH = "your_report_print_xpath"
        self.REFRESH_REPORT_XPATH = "your_refresh_report_xpath"
        self.EMAIL_TO_XPATH = "your_email_to_xpath"
        self.EMAIL_SEND_XPATH = "your_email_send_xpath"

        self.CHEQUE_REGISTER_XPATH = "your_cheque_register_xpath"
        self.CHEQUE_REGISTER_ADVANCE_SEARCH_XPATH = "your_cheque_register_advance_search_xpath"
        self.CHEQUE_REGISTER_DATE_XPATH = "your_cheque_register_date_xpath"

        self.BANK_STATEMENT_XPATH = "your_bank_statement_xpath"
        self.BANK_ACCOUNT_XPATH = "your_bank_account_xpath"
        self.BANK_STATEMENT_DATE_XPATH = "your_bank_statement_date_xpath"

        self.EMAIL_BUTTON_XPATH = "//*[contains(@onclick,'EmailAFMReports')]"

        # Dynamic value from your test setup
        self.carrier_name = "your_carrier_name"
        self.email_to_value = "test@example.com"
        self.currency_value_cad = "CAD"

    def _click(self, xpath):
        self.wait.until(EC.element_to_be_clickable((By.XPATH, xpath))).click()

    def _type(self, xpath, value):
        element = self.wait.until(EC.visibility_of_element_located((By.XPATH, xpath)))
        element.clear()
        element.send_keys(value)

    def _is_displayed(self, xpath):
        try:
            return self.wait.until(
                EC.visibility_of_element_located((By.XPATH, xpath))
            ).is_displayed()
        except:
            return False

    def _select_by_visible_text(self, xpath, text):
        dropdown = self.wait.until(EC.visibility_of_element_located((By.XPATH, xpath)))
        Select(dropdown).select_by_visible_text(text)

    def _capture_screen(self, name):
        self.driver.save_screenshot(f"reports/{name}.png")

    def _future_date(self, days=5):
        return (datetime.now() + timedelta(days=days)).strftime("%m/%d/%Y") + " 12:00 AM"

    def profit_and_loss_report(self):
        try:
            drop_future_date = self._future_date(5)
            self._click(self.REPORTS_XPATH)
            self._click(self.ACCOUNT_XPATH)
            self._click(self.PROFIT_LOSS_XPATH)
            self._click(self.PROFIT_ADVANCE_SEARCH_XPATH)
            self._select_by_visible_text(self.PROFIT_CURRENCY_XPATH, self.currency_value_cad)
            self._type(self.PROFIT_DATE_XPATH, drop_future_date)
            is_submit_button_displayed = self._is_displayed(self.SUBMIT_XPATH)
            self._click(self.SUBMIT_XPATH)
            return is_submit_button_displayed
        except Exception as e:
            print(f"profit_and_loss_report failed: {e}")
            return False

    def profit_and_loss_report_print(self):
        try:
            self._click(self.REPORT_PRINT_XPATH)
            is_refresh_button_displayed = self._is_displayed(self.REFRESH_REPORT_XPATH)
            self._click(self.REFRESH_REPORT_XPATH)
            self._capture_screen("profitAndLossReportPrint")
            return is_refresh_button_displayed
        except Exception as e:
            print(f"profit_and_loss_report_print failed: {e}")
            return False

    def profit_and_loss_report_email(self):
        try:
            email_btn = self.wait.until(
                EC.presence_of_element_located((By.XPATH, self.EMAIL_BUTTON_XPATH))
            )
            self.driver.execute_script("arguments[0].click();", email_btn)
            self._type(self.EMAIL_TO_XPATH, self.email_to_value)
            is_email_send_button_displayed = self._is_displayed(self.EMAIL_SEND_XPATH)
            self._click(self.EMAIL_SEND_XPATH)
            self._capture_screen("profitAndLossReportEmail")
            time.sleep(3)
            return is_email_send_button_displayed
        except Exception as e:
            print(f"profit_and_loss_report_email failed: {e}")
            return False

    def cheque_register_report(self):
        try:
            drop_future_date = self._future_date(5)
            self._click(self.ACCOUNT_XPATH)
            self._click(self.CHEQUE_REGISTER_XPATH)
            self._click(self.CHEQUE_REGISTER_ADVANCE_SEARCH_XPATH)
            self._type(self.CHEQUE_REGISTER_DATE_XPATH, drop_future_date)
            is_submit_button_displayed = self._is_displayed(self.SUBMIT_XPATH)
            self._click(self.SUBMIT_XPATH)
            return is_submit_button_displayed
        except Exception as e:
            print(f"cheque_register_report failed: {e}")
            return False

    def cheque_register_report_print(self):
        try:
            self._click(self.REPORT_PRINT_XPATH)
            is_refresh_button_displayed = self._is_displayed(self.REFRESH_REPORT_XPATH)
            self._click(self.REFRESH_REPORT_XPATH)
            self._capture_screen("chequeRegisterReportPrint")
            return is_refresh_button_displayed
        except Exception as e:
            print(f"cheque_register_report_print failed: {e}")
            return False

    def cheque_register_report_email(self):
        try:
            email_btn = self.wait.until(
                EC.presence_of_element_located((By.XPATH, self.EMAIL_BUTTON_XPATH))
            )
            self.driver.execute_script("arguments[0].click();", email_btn)
            self._type(self.EMAIL_TO_XPATH, self.email_to_value)
            is_email_send_button_displayed = self._is_displayed(self.EMAIL_SEND_XPATH)
            self._click(self.EMAIL_SEND_XPATH)
            self._capture_screen("chequeRegisterReportEmail")
            time.sleep(3)
            return is_email_send_button_displayed
        except Exception as e:
            print(f"cheque_register_report_email failed: {e}")
            return False

    def bank_statement_report(self):
        try:
            drop_future_date = self._future_date(5)
            self._click(self.ACCOUNT_XPATH)
            is_bank_statement_button_displayed = self._is_displayed(self.BANK_STATEMENT_XPATH)
            self._click(self.BANK_STATEMENT_XPATH)
            self._select_by_visible_text(self.BANK_ACCOUNT_XPATH, self.carrier_name)
            self._type(self.BANK_STATEMENT_DATE_XPATH, drop_future_date)
            is_submit_button_displayed = self._is_displayed(self.SUBMIT_XPATH)
            self._click(self.SUBMIT_XPATH)
            self._capture_screen("bankStatementReport")
            return is_bank_statement_button_displayed and is_submit_button_displayed
        except Exception as e:
            print(f"bank_statement_report failed: {e}")
            return False

    def bank_statement_report_print(self):
        try:
            self._click(self.REPORT_PRINT_XPATH)
            is_refresh_button_displayed = self._is_displayed(self.REFRESH_REPORT_XPATH)
            self._click(self.REFRESH_REPORT_XPATH)
            self._capture_screen("bankStatementReportPrint")
            return is_refresh_button_displayed
        except Exception as e:
            print(f"bank_statement_report_print failed: {e}")
            return False

    def bank_statement_report_email(self):
        try:
            email_btn = self.wait.until(
                EC.presence_of_element_located((By.XPATH, self.EMAIL_BUTTON_XPATH))
            )
            self.driver.execute_script("arguments[0].click();", email_btn)
            self._type(self.EMAIL_TO_XPATH, self.email_to_value)
            is_email_send_button_displayed = self._is_displayed(self.EMAIL_SEND_XPATH)
            self._click(self.EMAIL_SEND_XPATH)
            self._capture_screen("bankStatementReportEmail")
            time.sleep(3)
            return is_email_send_button_displayed
        except Exception as e:
            print(f"bank_statement_report_email failed: {e}")
            return False
